### PR TITLE
Support grpc Server & Client (WIP)

### PR DESCRIPTION
**Pull request by RyanKung**
## What does this pull request do?

Emm, Support grpc server.

## Related issues

https://github.com/elastic/apm-agent-python/issues/645

### CODE REVIEW

File: elasticapm/contrib/grpcio/__init__.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+        method = handler_call_details.method
+        elasticapm.set_transaction_name("gRPC %s" % method)
+        ret = self.wrap_response(response_future, method, tx)
+        return ret
+
+    def request_started(self, handler_call_details):
+        meta_data = handler_call_details.invocation_metadata[0]._asdict()
+        trace_parent = TraceParent.from_headers(meta_data)
+        method = handler_call_details.method
```
[the remark this answers is not in the document]
Comment (RyanKung) [reply elasticapm/contrib/grpcio/__init__.py]: fixed

File: elasticapm/contrib/grpcio/__init__.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+import grpc
+import elasticapm
+import elasticapm.instrumentation.control
+from elasticapm.base import Client
+from elasticapm.conf import constants, setup_logging
+from elasticapm.contrib.flask.utils import get_data_from_request, get_data_from_response
+from elasticapm.handlers.logging import LoggingHandler
+from elasticapm.traces import execution_context
+from elasticapm.utils import build_name_with_http_method_prefix
```
[the remark this answers is not in the document]
Comment (RyanKung) [reply elasticapm/contrib/grpcio/__init__.py]: fixed

File: elasticapm/contrib/grpcio/__init__.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+from __future__ import absolute_import
+
+import logging
+
+import grpc
+import elasticapm
+import elasticapm.instrumentation.control
+from elasticapm.base import Client
+from elasticapm.conf import constants, setup_logging
+from elasticapm.contrib.flask.utils import get_data_from_request, get_data_from_response
```
[the remark this answers is not in the document]
Comment (RyanKung) [reply elasticapm/contrib/grpcio/__init__.py]: fixed

File: elasticapm/contrib/grpcio/__init__.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+from __future__ import absolute_import
+
+import logging
+
+import grpc
+import elasticapm
+import elasticapm.instrumentation.control
+from elasticapm.base import Client
+from elasticapm.conf import constants, setup_logging
```
[the remark this answers is not in the document]
Comment (RyanKung) [reply elasticapm/contrib/grpcio/__init__.py]: fixed

**Comment by Zioyi** [comment]
I am care about this PR, how is going?[excited]

**Comment by skywalkerisnull** [comment]
I am also very interested in this PR. Is there anything we can do to get the ball rolling again @RyanKung @beniwohli

**Comment by yswtrue** [comment]
Hi, when will this RP merge?

**Comment by upcFrost** [comment]
Tried the code from this PR. My findings so far:
* won't work with `grpc.aio`, 
* won't work with `.future`
* it assumes that the traceparent is the first header in `handler_call_details.invocation_metadata[0]._asdict()`, which is not always the case. Also it doesn't seem to take the tracestate into account.

Asyncio and headers are easy to fix, but im not sure about the future.

**Comment by RyanKung** [comment]
Sorry guys, I don't have enough time to finish this PR.

**Pull request closed without merging**